packages/next/src/build/swc/index.ts

```diff
@@ -1130,12 +1133,61 @@ function bindingToApi(
     }
   }
 
+  /**
+   * Returns a new copy of next.js config object to avoid mutating the original.
+   *
+   * Also it does some augmentation to the configuration as well, for example set the
+   * turbopack's rules if `experimental.reactCompilerOptions` is set.
+   */
+  function augmentNextConfig(
+    originalNextConfig: NextConfigComplete,
+    projectPath: string
+  ): Record<string, any> {
+    let nextConfig = { ...(originalNextConfig as any) }
+
+    const reactCompilerOptions = nextConfig.experimental?.reactCompiler
+
+    // It is not easy to set the rules inside of rust as resolving, and passing the context identical to the webpack
+    // config is bit hard, also we can reuse same codes between webpack config in here.
+    if (reactCompilerOptions) {
+      if (!nextConfig.experimental.turbo) {
+        nextConfig.experimental.turbo = {}
+      }
+
+      if (!nextConfig.experimental.turbo.rules) {
+        nextConfig.experimental.turbo.rules = {}
+      }
+
+      // [TODO]: need to enable against '*.ts'
+      for (const key of ['*.ts', '*.js', '*.jsx', '*.tsx']) {
+        if (nextConfig.experimental.turbo.rules[key]) {
+          Log.warn(
+            `The React Compiler cannot be enabled automatically because 'experimental.turbo' contains a rule for '${key}'. Remove this rule, or add 'babel-loader' and 'babel-plugin-react-compiler' to the Turbopack configuration manually.`
+          )
+          break
```

Comment:
If we find a rule for `'*.jsx'`, we'd break here, but we've already modified it for `'*.ts', '*.js'`. so this'd leave the nextConfig in a weird state where we applied rules for `'*.ts', '*.js'` but nothing after that. it'd probably make more sense to check all the keys first, and then only modify it if none are present?

Reply:
ugh, yes, good point.